### PR TITLE
Fix SLURM node and processor stats calculation

### LINKED ISSUE #88: "209 of 186 Nodes Active" indicates a problem with system status app report for Pitzer Expansion
209 of 186 Nodes Active

screen 2020-09-15 at 8 38 02 PM

### PR DESCRIPTION
**Pull request by msquee**
* Fix nodes available and processors available calculations.

Closes #88.

**Screenshot:**
image

### CODE REVIEW

**Review by treydock — APPROVED** [review]

**Comment by ericfranz** [comment]
@treydock @msquee do you know why in this screenshot it says the total number of expansion nodes is 395 but we hae 12+340+42+4 == 398 nodes? Are the 3 missing nodes pulled for a reservation or some other use at this time?

**Comment by treydock** [comment]
Because the flag used in this code is showing allocated/idle but there are more node states than those 2 states, last I looked 2 nodes were maint (reserved) and 5 nodes were drained (offline).  

```
$ sinfo -a -h -N -o "%N|%T" | uniq | cut -d'|' -f2 | sort | uniq -c
    168 allocated
      5 drained
    145 idle
      2 maint
     80 mixed

$ sinfo -a -h -o '%A'
247/148

$ sinfo -a -h -o '%n,%A' | sort | wc -l
400
```

The first one comes out to 400 nodes, and last one is 1 node per line with alloc/idle item, also 400.  I'm guessing `%A` is losing nodes due to other states.

FWIW this is what I use in Prometheus:

```
$ sinfo -a -N -h -o "%N|%T"
```
You get one line per node per partition and since nodes often exist in multiple partitions, you have to build a list of "seen" nodes and only look at %N value once and discard duplicates.  This gives you the actual state but it would require this code to be aware of possible states which isn't that hard given they are all in the sinfo man page.  I'd just say "mix" or "allocated" is used, "idle" is idle and everything else is other which is what you get with this (but this is CPUs):

```
$ sinfo -a -o "%C"
CPUS(A/I/O/T)
9825/9119/240/19184
```

**Comment by treydock** [comment]
Also be aware that at this time there are 4 interactive nodes in SLURM but they are drained and there are 2 nodes from original pitzer that are reserved (maint). We have also moved 4 expansion nodes into a test environment so they aren't visible so the total normally for expansion is 402 (including interactive) and we've added 2 and then removed 4 so max at this time is 400.

**Comment by treydock** [comment]
Also @msquee be sure we are using `-a` flag otherwise some nodes could potentially be hidden if we ever used a hidden partition or used `AllowGroups` to limit access to some partitions.  For example right now the `systems` partition might not show up without the `-a` flag and that partition is currently all possible nodes.

**Comment by msquee** [comment]
Fixed, I added `%D` to `sinfo` and pull the total node count directly from there.
image

File: lib/slurm_squeue_client.rb [hunk 1]
```diff
@@ -80,6 +80,7 @@ def cluster_info
       procs_avail:    sinfo_out[1].to_i,
       nodes_used:     sinfo_out[4].to_i,
       nodes_idle:     sinfo_out[5].to_i,
```
Comment (treydock) [inline lib/slurm_squeue_client.rb:82]: procs_total is sinfo_out[3].

File: lib/slurm_squeue_client.rb [hunk 2]
```diff
@@ -120,20 +121,34 @@ def eligible_percent
...
+  #
+  # @return [Integer] the total number of idle/used jobs
+  def available_nodes
+    nodes_avail
+  end
+
+  # Total number of available and in use procs
+  #
+  # @return [Integer] the total number of procs 
+  def available_procs
```
Comment (treydock) [inline lib/slurm_squeue_client.rb:136]: Use the number reported by sinfo as this calculation missed "other" proc states.  The output from `%C` contains the total:

```
$ sinfo -a -o "%C"
CPUS(A/I/O/T)
9661/9283/240/19184
```

**Pull request merged**